### PR TITLE
Add a primitive lexer for Typst

### LINKED ISSUE #2558: Add lexer for Typst
[Typst](https://github.com/typst/typst) is a new markup language and I've noticed that `pygments` doesn't support it.

The syntax documentation can be found here:
https://typst.app/docs/reference/syntax/

For comparison, somebody has already implemented a grammar definition for Github Linguist:
https://github.com/michidk/typst-grammar/blob/master/grammars/tmlanguage.json

### PR DESCRIPTION
**Pull request by jvoisin**
This should close #2558.

I tried to add tests but:

```
jvoisin@home 20:49 (tyspt) ~/dev/pygments tox -- --update-goldens ~/text.typ 
Traceback (most recent call last):
  File "/usr/bin/tox", line 33, in <module>
    sys.exit(load_entry_point('tox==3.21.4', 'console_scripts', 'tox')())
  File "/usr/lib/python3/dist-packages/tox/session/__init__.py", line 44, in cmdline
    main(args)
  File "/usr/lib/python3/dist-packages/tox/session/__init__.py", line 65, in main
    config = load_config(args)
  File "/usr/lib/python3/dist-packages/tox/session/__init__.py", line 81, in load_config
    config = parseconfig(args)
  File "/usr/lib/python3/dist-packages/tox/config/__init__.py", line 280, in parseconfig
    ParseIni(config, config_file, content)
  File "/usr/lib/python3/dist-packages/tox/config/__init__.py", line 1276, in __init__
    raise tox.exception.ConfigError(
tox.exception.ConfigError: ConfigError: pylint failed with ConfigError: skip_install: boolean value "True # doesn't need installing Pygments into the venv" needs to be 'True' or 'False' at Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/tox/config/__init__.py", line 1252, in run
    results[name] = cur_self.make_envconfig(name, section, subs, config)
  File "/usr/lib/python3/dist-packages/tox/config/__init__.py", line 1407, in make_envconfig
    res = meth(env_attr.name, env_attr.default, replace=replace)
  File "/usr/lib/python3/dist-packages/tox/config/__init__.py", line 1696, in getbool
    raise tox.exception.ConfigError(
tox.exception.ConfigError: ConfigError: skip_install: boolean value "True # doesn't need installing Pygments into the venv" needs to be 'True' or 'False'

[1]
jvoisin@home 20:49 (tyspt) ~/dev/pygments 
```

Also, while the lexer is working ~ok, there is still a lot of space for improvements.

### CODE REVIEW

File: pygments/lexers/typst.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            (r'@[a-zA-Z_][a-zA-Z0-9_-]*', Name.Label),  # reference
+            (r'#(?:let|set|show)', Keyword.Declaration, 'inline_code'),
+            (r'(#[a-zA-Z_][a-zA-Z0-9_]*)(\[)', bygroups(Name.Function, Punctuation), 'root'),
+            (r'(#[a-zA-Z_][a-zA-Z0-9_]*)(\()', bygroups(Name.Function, Punctuation), 'inline_code'),
+            (r'#[a-zA-Z_][a-zA-Z0-9_]*', Name.Variable,),
+            (r'```(?:.|\n)*?```', String.Backtick),  # code block
+            (r'https?://[0-9a-zA-Z~/%#&=\',;\\.\\+\\?]*', Generic.Emph),  # links
+            (words(('---', '\\', '~', '--', '...'), suffix=r'\b'), Punctuation),
+            (r'\[', Punctuation, '#push'),
+            (r'\}|\]', Punctuation, '#pop'),
```
Comment (birkenfeld) [inline pygments/lexers/typst.py:54]: Why does only `[` push but `}` and `]` pop?

File: pygments/lexers/typst.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            (r'`[^`]*`', String.Backtick),  # inline code
+            (r'^\s*-', Punctuation),  # unnumbered list
+            (r'^\s*\+', Punctuation),  # numbered list
+            (r'^\s*[0-9.]+', Punctuation),  # numbered list variant
+            (r'^\s*/\s+[^:]+:', Punctuation),  # definition
+            (r'=>|<=|==|!=|>|<|-=|\+=|\*=|/=|\+|-|\\|\*', Operator),
+            (r'<[a-zA-Z_][a-zA-Z0-9_-]*>', Name.Label),  # label
+            (r'@[a-zA-Z_][a-zA-Z0-9_-]*', Name.Label),  # reference
+            (r'#(?:let|set|show)', Keyword.Declaration, 'inline_code'),
+            (r'(#[a-zA-Z_][a-zA-Z0-9_]*)(\[)', bygroups(Name.Function, Punctuation), 'root'),
```
Comment (birkenfeld) [inline pygments/lexers/typst.py:47]: Should this `#push` instead?

File: pygments/lexers/typst.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        ],
+        'code': [
+            include('comment'),
+            (r'and|or|not', Operator.Word),
+            (r'#(?:as|break|export|continue|else|for|if|import|in|include|return|while)\b', Keyword.Reserved),
+            (words(('auto', 'none', 'true', 'false'), suffix='\b'), Keyword.Constant),
+            (r'([0-9.]+)(mm|pt|cm|in|em|fr|%)', bygroups(Number, Keyword.Reserved)),
+            (words(('as', 'break', 'export', 'continue', 'else', 'for', 'if', 'import', 'in', 'include','return', 'while'), suffix='\b'), Keyword),
+            (words(('let', 'set', 'show'), suffix=r'\b'), Keyword.Declaration),
+            (r'"[^"]*"', String.Double),
```
Comment (birkenfeld) [inline pygments/lexers/typst.py:74]: No escape mechanism?

File: pygments/lexers/typst.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        'inline_code': [
+            (r'=\s*\{\n', Punctuation, 'code'),
+            (r'\(\n', Punctuation, 'code'),
+            (r';|\n(?=\])', Punctuation, '#pop'),
+            include('code'),
+        ],
+        'code': [
+            include('comment'),
+            (r'and|or|not', Operator.Word),
+            (r'#(?:as|break|export|continue|else|for|if|import|in|include|return|while)\b', Keyword.Reserved),
```
Comment (birkenfeld) [inline pygments/lexers/typst.py:69]: I guess I'd prefer `words` with an explicit `#` in them.

**Comment by jvoisin** [comment]
image

Getting there, but I can't manage to make pygments understand that inline code can become regular code via the open parenthesis :/

**Comment by jvoisin** [comment]
I've tried to add tests with a more recent version of tox, but:

```console
$ tox -- --update-goldens ~/cv_jvoisin.typ 
.pkg: _optional_hooks> python /home/jvoisin/dev/pygments/.venv/lib/python3.9/site-packages/pyproject_api/_backend.py True hatchling.build
.pkg: get_requires_for_build_editable> python /home/jvoisin/dev/pygments/.venv/lib/python3.9/site-packages/pyproject_api/_backend.py True hatchling.build
.pkg: build_editable> python /home/jvoisin/dev/pygments/.venv/lib/python3.9/site-packages/pyproject_api/_backend.py True hatchling.build
py: install_package> python -I -m pip install --force-reinstall --no-deps /home/jvoisin/dev/pygments/.tox/.tmp/package/5/pygments-2.17.1-py3-none-any.whl
py: commands[0]> pytest --update-goldens /home/jvoisin/test.typ
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --update-goldens
  inifile: None
  rootdir: /home/jvoisin

py: exit 4 (0.14 seconds) /home/jvoisin/dev/pygments> pytest --update-goldens /home/jvoisin/test.typ pid=27838
.pkg: _exit> python /home/jvoisin/dev/pygments/.venv/lib/python3.9/site-packages/pyproject_api/_backend.py True hatchling.build
  py: FAIL code 4 (0.59=setup[0.44]+cmd[0.14] seconds)
  evaluation failed :( (0.66 seconds)
[4]
$```

File: pygments/lexers/typst.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            (r'^\s*=+.*$', Generic.Heading),
+            (r'[*][^*]*[*]', Generic.Strong),
+            (r'_[^_]*_', Generic.Emph),
+            (r'\$', Punctuation, 'maths'),
+            (r'`[^`]*`', String.Backtick),  # inline code
+            (r'^\s*-', Punctuation),  # unnumbered list
+            (r'^\s*\+', Punctuation),  # numbered list
+            (r'^\s*[0-9.]+', Punctuation),  # numbered list variant
+            (r'^(\s*/\s+)([^:]+)(:)', bygroups(Punctuation, Name.Variable, Punctuation)),  # definitions
+            (r'=>|<=|==|!=|>|<|-=|\+=|\*=|/=|\+|-|\\|\*', Operator), # comparisons
```
[the remark this answers is not in the document]
Comment (jvoisin) [reply pygments/lexers/typst.py:50]: Indeed

File: pygments/lexers/typst.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    version_added = '2.18'
+
+    tokens = {
+        'root': [
+            include('markup'),
+        ],
+        'common': [
+            (r' |\t', Whitespace),
+            (r',', Punctuation),
+            (r'.', Text),
```
[the remark this answers is not in the document]
Comment (jvoisin) [reply pygments/lexers/typst.py]: This is horrible black magic :<

[the remark this answers is not in the document]
Comment (jeanas) [reply pygments/lexers/typst.py]: It's how most lexers work :shrug:

**Comment by jeanas** [comment]
I've applied a few fixes that I missed on the initial review.

The rule for `#include` isn't working, though. Try

```
#include "foo.typ"
```

in the example. AFAICS, `#include` is eaten as `Name.Variable` by the rule in the `'markup'` state, then `"foo.typ"` is outside code mode. Can it be fixed? What are the exact rules on code mode? (BTW, `#(...)` seems not to be working either.)

If you don't want to fix it now, we can always merge the PR without `#include`.

**Comment by jvoisin** [comment]
The rules are outlined here: https://typst.app/docs/reference/syntax/

Feel free to merge it as is, I'll try to improve it afterwards :)

**Comment by jeanas** [comment]
Thank you @jvoisin!

**Pull request merged**